src/wp-includes/webfonts-api/class-wp-webfonts-controller.php

```diff
@@ -0,0 +1,206 @@
+	 *
+	 * @var WP_Webfonts_Registry
+	 */
+	private $webfonts_registry;
+
+	/**
+	 * Instance of the provider's registry.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var WP_Webfonts_Provider_Registry
+	 */
+	private $providers;
+
+	/**
+	 * Stylesheet handle.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var string
+	 */
+	private $stylesheet_handle = '';
+
+	/**
+	 * Create the controller.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param WP_Webfonts_Registry          $webfonts_registry Instance of the webfonts registry.
+	 * @param WP_Webfonts_Provider_Registry $provider_registry Instance of the providers registry.
+	 */
+	public function __construct(
+		WP_Webfonts_Registry $webfonts_registry,
+		WP_Webfonts_Provider_Registry $provider_registry
+	) {
+		$this->webfonts_registry = $webfonts_registry;
+		$this->providers         = $provider_registry;
+	}
+
+	/**
+	 * Initializes the controller.
+	 *
+	 * @since 5.9.0
+	 */
+	public function init() {
+		$this->providers->init();
+
+		// Register callback to generate and enqueue styles.
+		if ( did_action( 'wp_enqueue_scripts' ) ) {
+			$this->stylesheet_handle = 'webfonts-footer';
+			$hook                    = 'wp_print_footer_scripts';
+		} else {
+			$this->stylesheet_handle = 'webfonts';
+			$hook                    = 'wp_enqueue_scripts';
+		}
+		add_action( $hook, array( $this, 'generate_and_enqueue_styles' ) );
+
+		// Enqueue webfonts in the block editor.
+		add_action( 'admin_init', array( $this, 'generate_and_enqueue_editor_styles' ) );
+
+		// Add resources hints.
+		add_filter( 'wp_resource_hints', array( $this, 'get_resource_hints' ), 10, 2 );
+	}
+
+	/**
+	 * Generate and enqueue webfonts styles.
+	 *
+	 * @since 5.9.0
+	 */
+	public function generate_and_enqueue_styles() {
+		// Generate the styles.
+		$styles = $this->generate_styles();
+
+		// Enqueue the stylesheet.
+		wp_register_style( $this->stylesheet_handle, '' );
+		wp_enqueue_style( $this->stylesheet_handle );
+
+		// Add the styles to the stylesheet.
+		wp_add_inline_style( $this->stylesheet_handle, $styles );
+	}
+
+	/**
+	 * Generate and enqueue editor styles.
+	 *
+	 * @since 5.9.0
+	 */
+	public function generate_and_enqueue_editor_styles() {
+		wp_add_inline_style( 'wp-block-library', $this->generate_styles() );
+	}
+
+	/**
+	 * Generate styles for webfonts.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return string $styles Generated styles.
+	 */
+	private function generate_styles() {
+		$styles    = '';
+		$providers = $this->get_providers()->get_all_registered();
+		foreach ( $providers as $provider_id => $provider ) {
+			$registered_webfonts = $this->webfonts_registry->get_by_provider( $provider_id );
+
+			if ( ! empty( $registered_webfonts ) && $provider->is_external() ) {
+				foreach ( $registered_webfonts as $key => $webfont ) {
+					if ( empty( $webfont['is-external'] ) || true !== $webfont['is-external'] ) {
+						unset( $registered_webfonts[ $key ] );
+					}
+				}
+			}
+
+			if ( empty( $registered_webfonts ) ) {
+				continue;
+			}
```

Comment:
Could be a micro-optimization, but I always like seeing the guard clause first (it's more 'natural' to me 😂)

```suggestion
			if ( empty( $registered_webfonts ) ) {
				continue;
			}
			
			if ( ! empty( $registered_webfonts ) && $provider->is_external() ) {
				foreach ( $registered_webfonts as $key => $webfont ) {
					if ( empty( $webfont['is-external'] ) || true !== $webfont['is-external'] ) {
						unset( $registered_webfonts[ $key ] );
					}
				}
			}
```

Also, if we do have that guard clause first, the`! empty( $registered_webfonts )` part of the second conditional seems unneeded.

Reply:
Notice that there's an `unset`. It's possible after running through it the array will be empty. Having the guard clause after further guards the code after it. What do you think @dingo-d?

Reply:
Yeah, the `unset` is why I added the clause there. But I'm curious if there's a better way to handle this!

Reply:
But if it's empty, how can you unset a key from it? 🤔

Reply:
@dingo-d Pushed an update https://github.com/WordPress/wordpress-develop/pull/1736/commits/0be2982061e4d38f067b9ccb49676b1194c08fa7. It accomplishes what you are seeking, i.e. bail out early, while also including more micro-optimizations and inline comments to explain what's going on. What do you think?

Reply:
Oh, this definitely is more readable and the comments are excellent 🎉 Thanks for taking my comments into the consideration 🙂